File: backend/app/models/fixed_asset_accounting.py
from decimal import Decimal
from typing import Literal

from app.models.accounting import JournalEntryRecord
from pydantic import BaseModel, ConfigDict, Field


FormalAssetAccountingStatus = Literal[
    "not_capitalized",
    "capitalized",
    "depreciating",
    "impaired",
    "disposed",
    "sold",
]
FormalAssetLifecycleStatus = Literal["active", "disposed", "sold"]
FixedAssetAccountingBatchStatus = Literal["generated", "existing", "skipped"]


class FormalAssetAccountingCard(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(min_length=1, max_length=64)
    asset_id: str = Field(min_length=1, max_length=80)
    asset_code: str = Field(min_length=1, max_length=80)
    asset_name: str = Field(min_length=1, max_length=120)
    category: str = Field(min_length=1, max_length=80)
    acquisition_date: str = Field(pattern=r"^\d{4}-\d{2}-\d{2}$")
    original_cost: Decimal = Field(ge=0, max_digits=14, decimal_places=2)
    salvage_value: Decimal = Field(ge=0, max_digits=14, decimal_places=2)
    useful_life_months: int = Field(ge=1, le=600)
    monthly_depreciation: Decimal = Field(ge=0, max_digits=14, decimal_places=2)
    accumulated_depreciation: Decimal = Field(ge=0, max_digits=14, decimal_places=2)
    impairment_amount: Decimal = Field(ge=0, max_digits=14, decimal_places=2)
    net_book_value: Decimal = Field(ge=0, max_digits=14, decimal_places=2)
    asset_status: FormalAssetLifecycleStatus
    formal_accounting_status: FormalAssetAccountingStatus
    capitalization_entry_id: str | None = None
    last_depreciation_entry_id: str | None = None
    last_depreciated_period: str | None = None
    impairment_entry_ids: list[str] = Field(default_factory=list)
    disposal_entry_ids: list[str] = Field(default_factory=list)


class FormalAssetAccountingCardListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str
    cards: list[FormalAssetAccountingCard]


class FixedAssetCapitalizationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(default="default", min_length=1, max_length=64)
    asset_id: str = Field(min_length=1, max_length=80)
    period: str = Field(pattern=r"^\d{4}-\d{2}$")
    credit_account_code: str = Field(default="2202", min_length=1, max_length=32)


class FixedAssetDepreciationPostRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(default="default", min_length=1, max_length=64)
    period: str = Field(pattern=r"^\d{4}-\d{2}$")


class FixedAssetImpairmentPostRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(default="default", min_length=1, max_length=64)
    asset_id: str = Field(min_length=1, max_length=80)
    period: str = Field(pattern=r"^\d{4}-\d{2}$")
    amount: Decimal = Field(gt=0, max_digits=14, decimal_places=2)


class FixedAssetDisposalPostRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(default="default", min_length=1, max_length=64)
    asset_id: str = Field(min_length=1, max_length=80)
    period: str = Field(pattern=r"^\d{4}-\d{2}$")
    proceeds_amount: Decimal = Field(ge=0, max_digits=14, decimal_places=2)
    disposal_date: str | None = Field(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$")
    proceeds_account_code: str = Field(default="1002", min_length=1, max_length=32)
    reason: str = Field(default="固定资产正式处置", min_length=1, max_length=200)


class FixedAssetAccountingEntryBatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str
    period: str
    status: FixedAssetAccountingBatchStatus
    depreciated_count: int = Field(ge=0)
    total_depreciation: Decimal = Field(ge=0, max_digits=14, decimal_places=2)
    entries: list[JournalEntryRecord]


class FixedAssetDisposalAccountingResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str
    period: str
    asset_id: str
    asset_code: str
    asset_status: FormalAssetLifecycleStatus
    clearing_account_code: str = "1606"
    disposal_gain_or_loss: Decimal = Field(max_digits=14, decimal_places=2)
    entries: list[JournalEntryRecord]

File: frontend/src/types/fixedAssetAccounting.ts
import type { JournalEntryRecord } from "./accounting";
import type { MoneyValue } from "./fixedAsset";

export type FormalAssetAccountingStatus =
  | "not_capitalized"
  | "capitalized"
  | "depreciating"
  | "impaired"
  | "disposed"
  | "sold"
  | string;

export interface FormalAssetAccountingCard {
  account_set_id: string;
  asset_id: string;
  asset_code: string;
  asset_name: string;
  category: string;
  acquisition_date: string;
  original_cost: MoneyValue;
  salvage_value: MoneyValue;
  useful_life_months: number;
  monthly_depreciation: MoneyValue;
  accumulated_depreciation: MoneyValue;
  impairment_amount: MoneyValue;
  net_book_value: MoneyValue;
  asset_status: string;
  formal_accounting_status: FormalAssetAccountingStatus;
  capitalization_entry_id: string | null;
  last_depreciation_entry_id: string | null;
  last_depreciated_period: string | null;
  impairment_entry_ids: string[];
  disposal_entry_ids: string[];
}

export interface FormalAssetAccountingCardListResponse {
  account_set_id: string;
  cards: FormalAssetAccountingCard[];
}

export interface FixedAssetCapitalizationRequest {
  account_set_id?: string;
  asset_id: string;
  period: string;
  credit_account_code?: string;
}

export interface FixedAssetDepreciationPostRequest {
  account_set_id?: string;
  period: string;
}

export interface FixedAssetImpairmentPostRequest {
  account_set_id?: string;
  asset_id: string;
  period: string;
  amount: MoneyValue;
}

export interface FixedAssetDisposalPostRequest {
  account_set_id?: string;
  asset_id: string;
  period: string;
  proceeds_amount: MoneyValue;
  disposal_date?: string | null;
  proceeds_account_code?: string;
  reason?: string;
}

export interface FixedAssetAccountingEntryBatch {
  account_set_id: string;
  period: string;
  status: "generated" | "existing" | "skipped" | string;
  depreciated_count: number;
  total_depreciation: MoneyValue;
  entries: JournalEntryRecord[];
}

export interface FixedAssetDisposalAccountingResult {
  account_set_id: string;
  period: string;
  asset_id: string;
  asset_code: string;
  asset_status: string;
  clearing_account_code: string;
  disposal_gain_or_loss: MoneyValue;
  entries: JournalEntryRecord[];
}
